### PR TITLE
[XN2vr37d] Add apache license headers

### PR DESCRIPTION
**Pull request by gem-neo4j**
This is built on top of [this PR ](https://github.com/neo4j/apoc/pull/369). APOC already declares Apache 2.0 as a license in the main part of the repo, but this makes it more in line with how our other products work and a more explicit way of doing licensing :)

### CODE REVIEW

**Review by nadja-muller — COMMENTED** [review]
Nice work! Left some questions and nitpicks

File: licenses/text/The Go license [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply licenses/text/The Go license:27]: Well we concatenate these files together to form the big license file, so probs don't want extra lines as it looks strange, and feels overkill to strip the text first, github warnings shouldn't be an issue for us :)

File: licenses/text/GNU General Public License, Version 3 [hunk 2]
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
+                       Version 3, 29 June 2007
```
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply licenses/text/GNU General Public License, Version 3:2]: Oh I took these files directly from GDS, but I realise APOC doesn't need this particular one so will just delete it instead :)

File: licenses/headers/NOTICE-header.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+However, if you have executed an End User Software License and Services
+Agreement or an OEM Software License and Support Services Agreement, or
+another commercial license agreement with Neo4j or one of its
+affiliates (each, a "Commercial Agreement"), the terms of the license in
+such Commercial Agreement will supersede the GPL and you may use the
+software solely pursuant to the terms of the relevant Commercial
+Agreement.
+
+Full license texts are found in LICENSES.txt.
```
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply licenses/headers/NOTICE-header.txt:22]: Yes :) That is the massive file that will get uploaded, it isn't committed in the repo yet (it just gets generated on build and is under build/reports). I need to add it into the Jar still (just haven't done that part yet!)

File: licenses/text/BSD-2-Clause [hunk 4]
```diff
@@ -0,0 +1,22 @@
+Copyright <year> <copyright holder>
```
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply licenses/text/BSD-2-Clause:1]: We can't as we show this as a generic license per group of libraries, this is what GDS does (search for this or [yyyy] on https://neo4j.com/terms/thirdpartylicenses/ to see :)

File: licenses/text/Apache-2.0 [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply licenses/text/Apache-2.0:189]: See above :) Nice question though!

File: licenses/headers/NOTICE-header.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply licenses/headers/NOTICE-header.txt:22]: Speaking of which, GDS doesn't commit the file to the repo, but I am wondering if that would be useful or not, what are your thoughts?

File: test-utils/build.gradle [hunk 6]
```diff
@@ -31,6 +31,8 @@ dependencies {
     api group: 'org.neo4j', name: 'log-test-utils', version: neo4jVersionEffective // , classifier: "tests"
     api group: 'org.neo4j.driver', name: 'neo4j-java-driver', version: '4.0.0'
     api group: 'org.apache.hadoop', name: 'hadoop-hdfs', version: '3.3.5', withoutServers.andThen(withoutLoggers)
+    // If updated check if the transitive dependency to javax.servlet.jsp:jsp-api:2.1 has also updated
+    // and remove the manual licensing check for it in licenses-3rdparties.gradle
```
[the remark this answers is not in the document]
Comment (gem-neo4j) [reply test-utils/build.gradle:35]: // If updated check if the transitive dependency to javax.servlet.jsp:jsp-api:2.1 has also updated
// and remove the manual licensing check for it in the downloadLicenses task in licenses-3rdparties.gradle

and then I can add the more specific instructions in that task

// If the dependency has been updated, remove the key from the map and try running ./gradlew generateLicensesFiles to verify the license is now discoverable (may need to do a ./gradlew clean to force an update to the dependency to license cached file!

Does this make sense? ^^

File: licenses/text/The Go license (hunk 1 above)
[the remark this answers is not in the document]
Comment (nadja-muller) [reply licenses/text/The Go license:27]: It looked liked some of them have a new line, though 😇

**Review by nadja-muller — APPROVED** [review]
Nice work 🙂. Not sure if it's worth looking over the new line no new line (non-)issue at the end of the license files.

**Pull request merged**